runners/direct-java/src/main/java/org/apache/beam/runners/direct/UnboundedReadEvaluatorFactory.java

```diff
@@ -147,29 +146,36 @@ public void processElement(
             // if the call to close throws an IOException.
             reader = null;
             toClose.close();
-            residual =
-                UnboundedSourceShard.of(
-                    shard.getSource(), shard.getDeduplicator(), null, finishedCheckpoint);
-          } else {
-            residual = shard.withCheckpoint(finishedCheckpoint);
           }
+          UnboundedSourceShard<OutputT, CheckpointMarkT> residual = UnboundedSourceShard.of(
```

Comment:
So the missing element (in the continuous input case) is that getReader is potentially creating a new reader which we don't populate in the shard, right?

The test we have demonstrates that with an input `UnboundedSourceShard` with a reader we'll reuse it, but I don't know if we have a test that demonstrates that with no input reader we'll produce a shard with one, which I think is the best way to describe the bug. Can you add such a test?

Reply:
In addition to the test I updated, right? Will do.

Reply:
> The test we have demonstrates that with an input UnboundedSourceShard with a reader we'll reuse it, but I don't know if we have a test that demonstrates that with no input reader we'll produce a shard with one, which I think is the best way to describe the bug.

Can you rephrase it? The test does start with a shard without a reader and ensures that it is reused.

Reply:
I mean it ensures that the reader is created only once. And similar to other tests in the file, it does not actually assert the output is correct.

Reply:
Basically, we have two cases of input:
`UnboundedSourceShard{source, deduplicator, reader.absent(), checkpoint}`
`UnboundedSourceShard{source, deduplicator, reader.present(), checkpoint}`

and we have a test to demonstrate that if given `UnboundedSourceShard{source, deduplicator, reader.present(), checkpoint}`, we produce `UnboundedSourceShard{source, deduplicator, reader.present(), newCheckpoint}`

but no test to demonstrate that if given `UnboundedSourceShard{source, deduplicator, reader.absent(), checkpoint}` we produce `UnboundedSourceShard{source, deduplicator, reader.present(), newCheckpoint}`, which is the base case that we don't handle properly right now, which means we almost never get to the reuse case.

Reply:
Line 291[1] create a `UnboundedSourceShard{source, deduplicator, reader.absent(), checkpoint}`, added to inputBundle right below that.

> and we have a test to demonstrate that if given UnboundedSourceShard{source, deduplicator, reader.present(), checkpoint}, we produce UnboundedSourceShard{source, deduplicator, reader.present(), newCheckpoint}

Are you referring to `evaluatorReusesReader()` updated here or is this another test?

Sorry for not multiple follow up questions, I think I am misinterpreting what you are saying.
 
[1]: https://github.com/apache/beam/pull/4658/files#diff-41695f8bd3f499370918186c7f369efdR291

Reply:
Ahah, yep, that's pretty busted;

Really I think the updates to `evaluatorReusesReader` is mostly checking that the reader is populated in the second shard.

We should have a test around reusing a reader that is then discarded because there was no input and it was at the end of time (the new branch you've added)

You can probably remove `withCheckpoint` from the `UnboundedSourceShard` class as well; I think this is probably the only caller.

Reply:
Removed `withCheckpoint`.
Added an option to advance  watermark to infinity when the reader reaches the end
Updated the test to:
  - verify that number of elements produced matches input.
  - reader is closed when end of stream is reached.